tls/s2n_cipher_suites.c

```diff
@@ -1227,15 +1228,27 @@ static int s2n_set_cipher_as_server(struct s2n_connection *conn, uint8_t *wire,
                 continue;
             }
 
+            /* For TLS1.3 when PSKs are present, the server must consider the hash algorithm associated with the chosen PSK
+             * when choosing a cipher suite. Server MUST reject any cipher suite without a matching hash algorithm and 
+             * continue to the next candidate. 
+             * */     
+            if (conn->psk_params.chosen_psk != NULL) {
+                s2n_hmac_algorithm chosen_psk_hmac_alg;
+                GUARD(s2n_hash_hmac_alg(conn->psk_params.chosen_psk->hash_alg, &chosen_psk_hmac_alg));
+                if (match->prf_alg != chosen_psk_hmac_alg) {
+                    continue;
+                }
+            }
+
             conn->secure.cipher_suite = match;
-            return 0;
+            return S2N_SUCCESS;
         }
     }
 
     /* Settle for a cipher with a higher required proto version, if it was set */
     if (higher_vers_match) {
         conn->secure.cipher_suite = higher_vers_match;
-        return 0;
+        return S2N_SUCCESS;
     }
```

Comment:
This creates a route where we could choose a cipher suite (higher_vers_match) that has a prf alg that doesn't match the PSK. You'll need to move your new check before higher_vers_match is set.

Reply:
I can move the check prior to setting the higher_vers_match, however this route would not be feasible as PSK hash match is required only for TLS1.3 versions. In the case of choosing a cipher suite using higher_vers_match the actual_protocol version would be <= TLS12 where PSKs are not supported and hence the PSK hash match not required.

Reply:
Please do update it. We shouldn't create a code path that's subtly only correct if TLS1.3 is the maximum protocol version, because theoretically someday TLS1.3 will not be the maximum protocol version :) If we're going to make assumptions like that, we need to be very explicit and intentional about them.